src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs

```diff
@@ -538,57 +539,67 @@ public void Reset()
                 cancellationToken);
         }
 
-        // These methods are for chunked http responses that use GetMemory/Advance
         private Memory<byte> GetChunkedMemory(int sizeHint)
         {
-            // The max size of a chunk will be the size of memory returned from the PipeWriter (today 4096)
-            // minus 5 for the max chunked prefix size and minus 2 for the chunked ending, leaving a total of
-            // 4089.
+            var originalSizeHint = sizeHint;
+
+            // To guarantee that sizeHint is a minimumSize, we need to call GetMemory with a sizeHint
+            // larger than what was provided. We add the begin and end chunk lengths here.
+            if (sizeHint > 0)
```

Comment:
What if the sizeHint is 0, and there is some space remaining in _currentChunkMemory, but not enough space for a chunk prefix/suffix plus real data?

I have the same question for when sizeHint is 0, there's no space remaining in _currentChunkMemory, and _pipeWriter.GetMemory(0) does not return enough space for a chunk?

Reply:
> What if the sizeHint is 0, and there is some space remaining in _currentChunkMemory, but not enough space for a chunk prefix/suffix plus real data?

The big thing to remember is when someone calls GetMemory and there is space remaining, we don't need to reinsert the chunk prefix/suffix. Here is an example.

Let's say we call GetMemory(0). This will call GetMemory on the PipeWriter, returning a chunk of 4096. We allocate space in the chunk like this:
[2 bytes chunked prefix | 3 bytes chunk length | 4089 bytes chunk data | 2 bytes chunked suffix]

When someone calls Advance(4088) (one less than the length of the chunk), the chunk will now look like:
[2 bytes chunked prefix | 3 bytes chunk length | 4088 bytes used chunk data | 1 byte unused chunk data | 2 bytes chunked suffix]

If someone calls GetMemory(0) it will return the 1 remaining byte in the chunk (without needing to insert the prefix/suffix) because we already wrapped the chunk. 

If someone calls GetMemory(2), we would write the current chunk before getting new memory.

> I have the same question for when sizeHint is 0, there's no space remaining in _currentChunkMemory, and _pipeWriter.GetMemory(0) does not return enough space for a chunk?

 Same applies as above, except the next call to GetMemory(0) would return a new chunk.